1056 GOcean transformations to insert boundary masks and eliminate return statements

sergisiso opened the pull request:
Closes #1056 by adding the following transformations:
- GOMoveIterationBoundariesInsideKernelTrans
- FoldConditionalReturnExpressionsTrans

This is in preparation of a kernel fusing transformation but also help in starting to move the gen_ocl_code() methods to transformations that apply to the PSyIR tree.

sergisiso commented:
This PR is ready for the first review from @arporter or @rupertford

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -1411,93 +1412,29 @@ def gen_arg_setter_code(self, parent):
-        # First the boundary values
-        for boundary in boundary_names:
+        # First the PSy-layer kernel arguments
+        for arg in self.arguments.args:
```
Could you fuse these two loops to reduce duplication? e.g.:

    for arg in [karg.name for karg in self.arguments.args] + boundary_names:

sergisiso replied:
Ah yes, this is much nicer.

arporter commented on src/psyclone/gocean1p0.py:
```diff
@@ -769,7 +769,7 @@ def _grid_property_psyir_expression(self, grid_property):
 
     # -------------------------------------------------------------------------
```
pylint complains about a few things imported but unused.

sergisiso replied:
Removed unused

arporter commented on src/psyclone/psyir/nodes/assignment.py:
```diff
@@ -197,3 +197,12 @@ def is_array_range(self):
+        :param parent: the parent of this Node in the PSyIR.
+        :type parent: :py:class:`psyclone.psyir.nodes.Node`
+        '''
+        from psyclone.f2pygen import PSyIRGen
```
If poss, please move this up.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+    (because happens at the end of the Routine). For example, the following
+    code:
+
+    >>> subroutine test(i)
```
Is it possible to tell Sphinx that this is Fortran?

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+class FoldConditionalReturnExpressionsTrans(Transformation):
+    ''' Provides a transformation that folds conditional expressions with only
+    a return statement inside so that the Return statement can disapear
+    (because happens at the end of the Routine). For example, the following
```
Could you add a sentence (before the "For example...") saying why one might want to do this?

sergisiso replied:
Done, also rephrased the the previous sentence of the docstring.

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+    '''
+
+    def __str__(self):
+        return "Fold all conditional expressions with Return statements."
```
Possibly re-word to make it clearer, e.g. "re-structure conditional expressions to eliminate early Return statements"?

sergisiso replied:
Updated description

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+        :param node: the node to validate.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
s/dictionary/dict/

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the node is not a Routine.
+        '''
```
Blank line inside docstring rather than after please.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+
+        if not isinstance(node, Routine):
+            raise TransformationError("Error in {0} transformation. "
+                                      "This transformation can only be applied"
```
Please include the type of the supplied node in the message.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+                 :py:class:`psyclone.undoredo.Memento`)
+        '''
+        routine = node
+        self.validate(routine)
```
Please include the options argument to `validate`.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+        :param node: the node to transform.
+        :type node: :py:class:`psyclone.psyir.nodes.Routine`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
s/dictionary/dict/

sergisiso replied:
Fixed

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+            :returns: whether the given node represents a conditional return \
+                      expression.
+            '''
+
```
Please rm this blank line.

sergisiso replied:
Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+                return False
+            return isinstance(node.if_body[0], Return)
+
+        remaining_statements = [line for line in routine]
```
pylint complains about this line. Could you just do `remaining_statements = routine[:]`?

sergisiso replied:
Removed line

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+                    UnaryOperation.Operator.NOT,
+                    statement.condition)
+                statement.children[0] = new_condition
+                new_condition.parent = statement
```
Does `UnaryOperation.create()` allow you to specify `parent`?

sergisiso replied:
I doesn't at the moment.
More than having it as a create() parameter, ideally, I would like for the `statement.children[0] = new_condition` ,which is (I think) already intercepted by our custom Node::ChildrenList, to manage the parent link updates transparently.

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+                statement.children[0] = new_condition
+                new_condition.parent = statement
+
+                # Remove return (and dead code after return) and add remaining
```
Do you mean you're allowing for the case that the original code has some unreachable statements, e.g.:

    return
    not = reachable

If so, I suggest "and any dead code after the return".

sergisiso replied:
That's right. Done

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+
+                # Remove return (and dead code after return) and add remaining
+                # part of the routine schedule inside the loop body
+                for i in reversed(range(len(statement.if_body.children))):
```
Maybe:

    while statement.if_body.children:
        del statement.if_body.children[-1]

Possibly, with a comment explaining why we can't just do `statement.if_body.children = []`?

sergisiso replied:
Apparently we can do `statement.if_body.children = []`, I was overcomplicating things.

arporter replied:
Is this OK though? I thought `children` had to be an instance of `ChildrenList` in order to get the validation checks? I suppose we could just create a new `ChildrenList` instead of `[]`?

sergisiso replied:
Yes it is, the assignment is a setter property that converts the provided list to ChildrenList. This is why things like `assingnment.children = [lhs, rhs]` are validated :)

arporter commented on src/psyclone/psyir/transformations/fold_conditional_return_expressions_trans.py:
```diff
@@ -0,0 +1,148 @@
+                start = statement.position
+                end = len(statement.parent.children) - 1
+                for index in range(end, start, -1):
+                    move = statement.parent.children[index]
```
I think you could do `move = statement.parent.children.pop(index)`

sergisiso replied:
Good idea. Done

sergisiso replied:
In fact since pop() already picks the last one by default I could get rid of index and the range() as well.

arporter commented on src/psyclone/tests/domain/gocean/goloop_test.py:
```diff
@@ -133,10 +133,10 @@ def test_goloop_bounds_invalid_iteration_space():
     gojloop._iteration_space = "broken"
     with pytest.raises(GenerationError) as err:
-        gojloop._upper_bound()
+        gojloop.upper_bound()
```
Copyright date?

sergisiso replied:
I think it is already up-to-date (maybe by a merge?)

arporter commented on src/psyclone/tests/domain/gocean/transformations/gocean1p0_transformations_test.py:
```diff
@@ -1527,6 +1529,12 @@ def test_ocl_apply(kernel_outputdir):
                              "one_invoke.f90", API, idx=0, dist_mem=False)
     schedule = invoke.schedule
+    # Currently, moving the boundaries inside the kernel is a prerequisite
+    # for OCLTrans
```
We need to extend the validation in OCLTrans to check that this has been done.

sergisiso replied:
My comments here were not accurate. It is in not a prerequisite of OCLTrans (the transformations can be applied in any order) so the validation would not help.

What it is at the moment is a prerequisite for the gen_ocl() done during the generation step, not doing it would already fail with a generation error that now mention this transformation, see gocean1p0_opencl_test.py::test_opencl_kernel_missing_boundary_symbol

Note that I expect this limitation to be short lived as hopefully #1134 will make this more flexible.

arporter reviewed (changes requested):
Looks good Sergi - it's very encouraging to see how easy it is to manipulate the PSyIR.
Please could you update the User Guide with the new transformations.
I think we're currently missing some validation - the OCLTrans needs to check that the loop-bounds-into-kernels transformation has been done for all kernels in the schedule.
I've looked at all source files and a couple of the test files so far and have checked with pylint/pycodestyle as I went.
All tests pass with compilation enabled.

arporter commented:
While looking at some other code I found how you tell Sphinx about Fortran in a doc string:

    .. code-block:: fortran

        DO i=1, n

sergisiso commented on examples/gocean/eg1/opencl_transformation.py:
```diff
@@ -69,6 +73,7 @@ def trans(psy):
         # Remove the globals from inside each kernel
         for kern in schedule.kernels():
             print("Remove globals from kernel: " + kern.name)
+            move_boundaries_trans.apply(kern)
```
Done

sergisiso commented on examples/gocean/eg3/ocl_trans.py:
```diff
@@ -48,18 +54,21 @@ def trans(psy):
 
     # Provide kernel-specific OpenCL optimization options
     for kern in sched.kernels():
+        move_boundaries_trans.apply(kern)
```
Added the comments, I slightly modified the README but not much because it doesn't go into details about the specific transformations and I think the overall descriptions still fit well.

sergisiso commented on examples/gocean/eg4/ocl_transform.py:
```diff
@@ -65,7 +68,9 @@ def trans(psy):
     # Provide kernel-specific OpenCL optimization options
+    move_boundaries_trans = GOMoveIterationBoundariesInsideKernelTrans()
     for kern in sched.kernels():
+        move_boundaries_trans.apply(kern)
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+    >>> stopy = N - 1
+    >>> do i = 1, size(field, 1)
+    >>>   do j = 1, size(field, 2)
+    >>>      kernel(i, j, field, startx, stopx, starty, stopx)
```
Fixed

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+
+    For example the following kernel call:
+
+    >>> do i = 2, N - 1
```
Updated to fortran code-block

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+    additionally a mask like the following one will be introduced in the
+    kernel code:
+
+    >>> if (i < startx .or. i > stopx .or. j < starty .or. j > stopx) then
```
Fixed

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        :type options: dictionary of string:values or None
+
+        :raises TransformationError: if the node is not a CodedKern.
+         '''
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        :param node: the node to validate.
+        :type node: :py:class:`psyclone.psyGen.CodedKern`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
Updated

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        :param node: the node to transform.
+        :type node: :py:class:`psyclone.psyGen.CodedKern`
+        :param options: a dictionary with options for transformations.
+        :type options: dictionary of string:values or None
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        if not isinstance(node, CodedKern):
+            raise TransformationError("Error in {0} transformation. "
+                                      "This transformation can only be applied"
+                                      " to CodedKern nodes.".format(self.name))
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        :raises TransformationError: if the node is not a CodedKern.
+         '''
+
+        if not isinstance(node, CodedKern):
```
Agreed, it's better to have tighter checks .

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+                 :py:class:`psyclone.undoredo.Memento`)
+
+        '''
+        self.validate(node)
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+
+        # Get useful references
+        invoke_st = node.ancestor(InvokeSchedule).symbol_table
+        inner_loop = node.parent.parent
```
Done

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        outer_loop = inner_loop.parent.parent
+        cursor = outer_loop.position
+
+        # Create new symbols in the PSylayer and initialise them with
```
Ops. Finished now.

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+            raw_arguments.append(symbol.name)
+
+        # Now that the boundaries are inside the kernel, the looping should go
+        # trough all the field points
```
Fixed

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+            iteration_indices + data_arguments +
+            [xstart_symbol, xstop_symbol, ystart_symbol, ystop_symbol])
+
+        # Create boundaries masking condition
```
Fixed

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+                condition4)
+            )
+
+        # Insert if condition masking as the kernel first statement
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1411,93 +1412,29 @@ def gen_arg_setter_code(self, parent):
-                condition4)
-            )
-
-        # Insert if condition masking as the kernel first statement
```
That's right, this code modification is now a transformation that can be applied anytime.

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+        iteration_indices = kernel_st.iteration_indices
+        data_arguments = kernel_st.data_arguments
+
+        # Create new symbols and insert them as kernel arguments after
```
Maybe it can be done but then a whole grid or field structure (without the data) should be passed to the kernel, and it is not a good idea to pass big arguments to device kernels, so in this case I prefer to keep passing just the scalars for each boundary.

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+                                    outer_loop.upper_bound())
+        outer_loop.parent.children.insert(cursor, assign4)
+
+        # Update Kernel Call (only works with the gen_ocl_code)
```
I was being lazy and only working with my use case, I now update the Arguments.args instead of the raw_args_list() and it should work in every scenario. I added a new test showing that it also work as expected with the Psy-layer produced by the PSyIR Fortran Writer :)

sergisiso commented on src/psyclone/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans.py:
```diff
@@ -0,0 +1,226 @@
+
+        # Update Kernel Call (only works with the gen_ocl_code)
+        raw_arguments = node.arguments.raw_arg_list()
+        for symbol in [inv_xstart, inv_xstop, inv_ystart, inv_ystop]:
```
Now it uses the Arguments class (see previous comment), which doesn't have an extend() method, so I will leave the for construct for now.

sergisiso commented:
@arporter This is ready for the next review. I addressed your comments and add the transformations into the User Guide.

Note that I now support the transformations with and without OpenCL and using the gen_code/ocl() and the Fortran Backend Writer which removes the main limitation this had in the first review.

I also improved the situation of #798 (gen_ocl() duplicate variables) which appeared in some tests but anyway I think this code should go away soon.

arporter commented on src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py:
```diff
@@ -0,0 +1,162 @@
+    "  end if\n\n"
+    "end subroutine sub1\n")
+
+# Tests with preceding code that is not part of the mask
```
Please could you extend this comment to say that a mask is only identified as such if the first statement within the body of the If is a `Return` - anything else is ignored by this transformation. (I say this because it took me a while to remember what's going on.)

sergisiso replied:
Improved the comment, let me know if now it captures better what is being tested.

arporter commented on src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py:
```diff
@@ -0,0 +1,162 @@
+    "  a=0.0\n"
+    "end subroutine\n")
+SUB_IN2_1 = SUB_IN2.format("a = 0.0")
+SUB_IN2_2 = SUB_IN2.format(
```
Please add comments in here detailing why each case will be ignored by the transformation.

sergisiso replied:
Updated the comment at the top. Is that enough?

arporter replied:
Yes, I think it's clear now, thanks.

arporter commented on src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py:
```diff
@@ -0,0 +1,162 @@
+              (SUB_IN2_2, SUB_OUT2_2), (SUB_IN2_3, SUB_OUT2_3)]
+
+
+@pytest.mark.parametrize("test_case", [0, 1, 2, 3])
```
You could do:

    @pytest.mark.parametrize("input_code, expected", [(SUB_IN1, SUB_OUT1), (....), ...]
    def test_transformation(parser, input_code, expected):

sergisiso replied:
I had this before but then I realised that for failed tests or running "pytest -v" it was including the whole source code (without line breaks) as the name of the test. Now it looks like this:

```
=========================================================================================== test session starts ===========================================================================================
platform linux -- Python 3.8.5, pytest-6.0.1, py-1.9.0, pluggy-0.13.1 -- /home/sergi/.virtualenvs/psy3/bin/python
cachedir: .pytest_cache
rootdir: /home/sergi/workspace/stfc/PSyclone, configfile: setup.cfg
plugins: forked-1.3.0, flakes-4.0.1, pylint-0.17.0, xdist-2.1.0, cov-2.10.1, pep8-1.0.6, pep257-0.0.5
collected 6 items

src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_description PASSED                                                                                 [ 16%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_validation PASSED                                                                                  [ 33%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[0] PASSED                                                                           [ 50%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[1] PASSED                                                                           [ 66%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[2] PASSED                                                                           [ 83%]
src/psyclone/tests/psyir/transformations/fold_conditional_return_expressions_trans_test.py::test_transformation[3] PASSED                                                                           [100%]
```
I though this was cleaner, but I can reverse it if you prefer.

arporter replied:
Oh I see. I don't mind much either way but I agree this output is cleaner so am happy to stick with it.

arporter reviewed (changes requested):
Thanks for generalising the kernel-argument handling.
I've now checked all files, including pylint.
All examples and tests are OK with compilation enabled.
Documentation has been updated and builds fine. Please could you check that you've populated `__all__` lists with the new classes added in this PR.
Please see inline for further comments.

sergisiso commented on doc/user_guide/transformations.rst:
```diff
@@ -425,14 +413,36 @@ variable that is available to it from the enclosing module scope.
+kernel. There are a number of transformations not listed here that can be
+applied to either or both the PSy-layer and Kernel-layer PSyIR.
+
+.. note:: Some of this transformation modifies the PSyIR tree of both: the
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -285,6 +284,33 @@ def gen_code(self, parent):
         :type parent: :py:class:`psyclone.f2pygen.ModuleGen`
         '''
+
+        def split_by_interface(symbol_names):
```
Good idea, changed to "sort_by_interface"

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -285,6 +284,33 @@ def gen_code(self, parent):
+
+        def split_by_interface(symbol_names):
+            ''' Utility to help differentiate between arguments, local symbols
+            and other symbols (globals, unresolved, ...)
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -285,6 +284,33 @@ def gen_code(self, parent):
+            :returns: a tuple of 3 lists with the given symbol_names \
+                      categorised as argument symbols, local symbols and \
+                      other symbols.
+            :rtype: 3-tuple
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -930,6 +952,19 @@ def lower_bound(self):
+        '''
+        In-place replacement of DSL or high-level concepts into generic
+        PSyIR constructs. A GOLoop needs to make sure the start and stop
+        expression of the Loop contain the boundaries defined by the API.
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1360,47 +1388,68 @@ def gen_arg_setter_code(self, parent):
                 if x.is_scalar and x.intrinsic_type == "integer"]
         if args:
+            # Use symbol table to avoid name clashes
+            names = []
```
Done

sergisiso commented on src/psyclone/gocean1p0.py:
```diff
@@ -1348,7 +1376,7 @@ def gen_arg_setter_code(self, parent):
         argsetter_st.add(RoutineSymbol(sub_name), tag=sub_name)
 
-        sub = SubroutineGen(parent, name=sub_name, args=args)
+        sub = SubroutineGen(parent, name=sub_name)
```
Done

sergisiso commented on src/psyclone/tests/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans_test.py:
```diff
@@ -137,3 +150,71 @@ def test_go_move_iteration_boundaries_inside_kernel_trans():
+
+def test_go_move_iteration_boundaries_inside_kernel_two_kernels_apply_twice():
+    ''' Tests that the GOMoveIterationBoundariesInsideKernelTrans
+    transformation for the GOcean API produces the expect code when the invoke
```
Fixed

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -638,21 +654,18 @@ def test_set_kern_args_real_grid_property():
-    # for the purpose of this test that just checks that the grid property
-    # dx is declared as 'REAL(KIND=go_wp), intent(in), target :: dx'
-    # assert GOcean1p0OpenCLBuild(kernel_outputdir).code_compiles(psy)
+    # This example can not be compile because it needs to import the
```
Fixed test. I found that issue #284 already describes this problem, added a TODO.

sergisiso commented on src/psyclone/tests/domain/gocean/transformations/gocean_move_iteration_boundaries_inside_kernel_trans_test.py:
```diff
@@ -0,0 +1,220 @@
+# Author S. Siso, STFC Daresbury Lab
+# -----------------------------------------------------------------------------
+
+''' Module containing tests for PSyclone
```
Done

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -374,6 +425,12 @@ def test_psy_init_with_options(kernel_outputdir):
     psy, _ = get_invoke("single_invoke.f90", API, idx=0)
     sched = psy.invokes.invoke_list[0].schedule
+    # Currently, moving the boundaries inside the kernel is a prerequisite
+    # for the GOcean gen_ocl() code generation.
```
I thought about this but ideally I would want it to be a optional transformation. I hope #1134 will make this possible, I will address it soon but its a complex PR. Probably I can re-evaluate the final decision about this as part of that PR (I will add a note in #1134).

sergisiso commented on src/psyclone/tests/gocean1p0_opencl_test.py:
```diff
@@ -504,50 +581,60 @@ def test_set_kern_args(kernel_outputdir):
-    assert generated_code.count("SUBROUTINE time_smooth_code_set_args("
-                                "kernel_obj, xstart, xstop, ystart, ystop, "
-                                "u_fld, unew_fld, uold_fld)") == 1
+
```
It works, of course it needs to add the new transformation into the script. I pushed https://github.com/stfc/PSycloneBench/pull/66 to make the change when this PR is merged.

sergisiso commented:
@arporter Ready for next review

arporter reviewed (changes requested):
Pretty much there now.
Please could you bring up-to-date with master and add `__all__` lists for any new/touched classes.

sergisiso commented:
@arporter Addressed your comments, ready for next review

arporter reviewed (approved):
Requested changes have been made.
Reference guide builds cleanly.
All examples and tests are still OK.
